### PR TITLE
.tools: Add per-commit Go checks (vet, fmt, and lint)

### PR DESCRIPTION
**Pull request by wking**
Instead of just checking the tip, run the checks on each commit in the
series.  This makes sure we catch errors with earlier commits even if
they were fixed by subsequent commits.

There are some details about low-level choices in the commit messages
themselves, but nothing that seems generic enough to be worth copying
out to the PR level.

This PR is based on #174, because as the output gets more verbose
(e.g. YAML blocks for golint's stdout) you'll likely want to feed it
into a TAP harness for compaction.  The non-verbose output is still
pretty short, although there are now more tests which you can fail ;).

If the protobuf branch lands (#179), we can swap out the Go checkers
and swap in [protoc-gen-lint][1].  I'm not aware of any protobuf
equivalent to 'go fmt' though, which is unfortunate.

[1]: https://github.com/ckaznocha/protoc-gen-lint

### CODE REVIEW

**Comment by wking** [comment]
I added a `git show --check …` too and some `TRAVIS_COMMIT_RANGE` debugging.  It turns out that the extra commits are because `TRAVIS_COMMIT_RANGE` is `{target-tip}...{pr-tip}` when it should be `{target-tip}..{pr-tip}`.  There shouldn't be a need to test commits that are only reachable from `{target-tip}`.  I'll see if I can find a Travis bug for this…

**Comment by wking** [comment]
> `TRAVIS_COMMIT_RANGE` is `{target-tip}...{pr-tip}` when it should be `{target-tip}..{pr-tip}`.

This is travis-ci/travis-ci#4596.  Until that gets fixed, I've pushed a workaround with 8831b20.

**Comment by vbatts** [comment]
honestly. Let's keep the TAP out for now.

**Comment by wking** [comment]
On Mon, Oct 05, 2015 at 12:16:37PM -0700, Vincent Batts wrote:
> honestly. Let's keep the TAP out for now. 

Ok (although I don't see a need to invent a different output syntax).
What syntax would you like for pass/fail (where TAP uses "ok" and "not
ok")?

File: .tools/validate.go [hunk 1]
```diff
@@ -39,6 +43,35 @@ var DefaultRules = []ValidateRule{
...
+		return vr
+	},
+	func(c CommitEntry) (vr ValidateResult) {
+		return ExecTree(c, "go", "vet", "./...")
+	},
+	func(c CommitEntry) (vr ValidateResult) {
+		return ExecTree(c, "go", "fmt", "./...")
+	},
+	func(c CommitEntry) (vr ValidateResult) {
+		vr = ExecTree(c, os.ExpandEnv("$HOME/gopath/bin/golint"), "./...")
```
Comment (vbatts) [inline .tools/validate.go]: but this is what we're doing in travis, and seems cleaner there.

**Comment by vbatts** [comment]
Honestly, that is not something I see a big need for a change and would be
fine with `$?` being not zero for failure, but it already prints out a
pass/fail as well.

File: .tools/validate.go (hunk 1 above)
Comment (wking) [inline .tools/validate.go]: On Mon, Oct 05, 2015 at 12:20:58PM -0700, Vincent Batts wrote:
> > +		vr = ExecTree(c, os.ExpandEnv("$HOME/gopath/bin/golint"), "./...")
> 
> but this is what we're doing in travis, and seems cleaner there.

You're not doing it for every commit in Travis, so you'd miss cases
where one commit breaks the format and a subsequent commit fixed it.

Comment (vbatts) [inline .tools/validate.go]: fair, but not completely a concern right now.

**Comment by wking** [comment]
On Mon, Oct 05, 2015 at 12:26:10PM -0700, Vincent Batts wrote:
> Honestly, that is not something I see a big need for a change and
> would be fine with `$?` being not zero for failure, but it already
> prints out a pass/fail as well.

It doesn't currently print pass/fail for the individual checks, it
just prints all the checks that passed [1] or all the checks that
failed [2].  There's no code currently for generating output like:

  FAIL
  ok 1 - has a valid DCO
  not ok 2 - git show --check 8831b203606aa74d2bf5274108816dde7629fa6f
  …

where a single commit has both passing and failing checks, and we
still see the passing checks.

[1]: https://github.com/opencontainers/specs/blob/f75f23fe4e9ad3cf1324530ee8876b1b3d2c8237/.tools/validate.go#L80
[2]: https://github.com/opencontainers/specs/blob/f75f23fe4e9ad3cf1324530ee8876b1b3d2c8237/.tools/validate.go#L89

**Comment by crosbymichael** [comment]
This was moved to a separate project out of this repo.

**Pull request closed without merging**